tests/forloops.test.ts

```diff
@@ -0,0 +1,339 @@
+   `10`,
+   `3`,
+   `10`]);
+  // 7
+  assertTCFail("for-loop changing iterator type inside loop", `
+  class Range(object):
+  current : int = 0
+  min : int = 0
+  max : int = 0
+  def new(self:Range, min:int, max:int)->Range:
+    self.min = min
+    self.current = min
+    self.max = max
+    return self
+  def next(self:Range)->int:
+    c : int = 0
+    c = self.current
+    self.current = self.current + 1
+    return c
+  def hasnext(self:Range)->bool:
+    return self.current < self.max
+
+
+
+
+cls:Range = None
+i:int = 0
+cls = Range().new(1, 4)
+ 
+for i in cls:
+   print(i)
+   i = True
+   print(i)` );
+
+   // 8
+  assertPrint("for-loop continue", `
+  class Range(object):
+  current : int = 0
+  min : int = 0
+  max : int = 0
+  def new(self:Range, min:int, max:int)->Range:
+    self.min = min
+    self.current = min
+    self.max = max
+    return self
+  def next(self:Range)->int:
+    c : int = 0
+    c = self.current
+    self.current = self.current + 1
+    return c
+  def hasnext(self:Range)->bool:
+    return self.current < self.max
+
+
+
+
+cls:Range = None
+i:int = 0
+cls = Range().new(1, 4)
+ 
+for i in cls:
+   print(i)
+   continue 
+   print(i)
+   
+i = 20
+print(i)`, [`1`, 
+   `2`,
+   `3`,
+   `20`]);
+
+ // 9
+ assertPrint("for-loop break", `
+ class Range(object):
+ current : int = 0
+ min : int = 0
+ max : int = 0
+ def new(self:Range, min:int, max:int)->Range:
+   self.min = min
+   self.current = min
+   self.max = max
+   return self
+ def next(self:Range)->int:
+   c : int = 0
+   c = self.current
+   self.current = self.current + 1
+   return c
+ def hasnext(self:Range)->bool:
+   return self.current < self.max
+
+
+
+
+cls:Range = None
+i:int = 0
+cls = Range().new(1, 4)
+
+for i in cls:
+  print(i)
+  break 
+  print(i)
+  
+i = 20
+print(i)`, [`1`, 
+  `20`]);
+ //10
+  assertPrint("for-loop function", `
+  class Range(object):
+  current : int = 0
+  min : int = 0
+  max : int = 0
+  def new(self:Range, min:int, max:int)->Range:
+    self.min = min
+    self.current = min
+    self.max = max
+    return self
+  def next(self:Range)->int:
+    c : int = 0
+    c = self.current
+    self.current = self.current + 1
+    return c
+  def hasnext(self:Range)->bool:
+    return self.current < self.max
+ 
+def checkForLoop(n:int):
+  c:Range = None
+  i:int = 0
+  c = Range().new(n,n+5)
+  for i in c:
+    print(i)
+  return
+
+checkForLoop(2)
+`, [`2`, 
+   `3`,
+  `4`,
+   `5`,
+  `6`]);
+  //11
+  assertTCFail("for-loop non id iterator", `class Range(object):
+  current : int = 0
+  min : int = 0
+  max : int = 0
+  def new(self:Range, min:int, max:int)->Range:
+    self.min = min
+    self.current = min
+    self.max = max
+    return self
+  def next1(self:Range)->int:
+    c : int = 0
+    c = self.current
+    self.current = self.current + 1
+    return c
+  def hasnext(self:Range)->bool:
+    return self.current < self.max
+
+
+
+
+cls:Range = None
+i:int = 0
+cls = Range().new(1,3)
+for 1+2 in cls:
```

Comment:
I'm genuinely surprised this is a type error and not a parse error. Is our testing util doing what we expect?

Reply:
The implementation for week-7 considers iterator as an expression where 1+2 is valid. A type check on iterator to be a "id" is performed in type-check.ts due to which this testcase results in type error instead of parse error.  

The iterator in week-8 implementation is considered as string and the corresponding checks are added in the parser.